solr/core/src/java/org/apache/solr/servlet/RateLimitManager.java

```diff
@@ -0,0 +1,148 @@
+ * rate limiting is being done for a specific request type.
+ */
+public class RateLimitManager {
+  public final static int DEFAULT_CONCURRENT_REQUESTS= (Runtime.getRuntime().availableProcessors()) * 3;
+  public final static long DEFAULT_SLOT_ACQUISITION_TIMEOUT_MS = -1;
+  private final Map<String, RequestRateLimiter> requestRateLimiterMap;
+  private final Map<HttpServletRequest, RequestRateLimiter> activeRequestsMap;
+
+  public RateLimitManager() {
+    this.requestRateLimiterMap = new HashMap<>();
+    this.activeRequestsMap = new ConcurrentHashMap<>();
+  }
+
+  // Handles an incoming request. The main orchestration code path, this method will
+  // identify which (if any) rate limiter can handle this request. Internal requests will not be
+  // rate limited
+  // Returns true if request is accepted for processing, false if it should be rejected
+  public boolean handleRequest(HttpServletRequest request) throws InterruptedException {
+    String requestContext = request.getHeader(SOLR_REQUEST_CONTEXT_PARAM);
+    String typeOfRequest = request.getHeader(SOLR_REQUEST_TYPE_PARAM);
+
+    if (typeOfRequest == null) {
+      // Cannot determine if this request should be throttled
+      return true;
+    }
+
+    // Do not throttle internal requests
+    if (requestContext != null && requestContext.equals(SolrRequest.SolrClientContext.SERVER.toString())) {
+      return true;
+    }
+
+    RequestRateLimiter requestRateLimiter = requestRateLimiterMap.get(typeOfRequest);
+
+    if (requestRateLimiter == null) {
+      // No request rate limiter for this request type
+      return true;
+    }
+
+    if (requestRateLimiter.handleRequest()) {
+      activeRequestsMap.put(request, requestRateLimiter);
+      return true;
+    }
+
+    requestRateLimiter = trySlotBorrowing(typeOfRequest);
+
+    if (requestRateLimiter != null) {
+      activeRequestsMap.put(request, requestRateLimiter);
+      return true;
+    }
+
+    return false;
+  }
+
+  /* For a rejected request type, do the following:
+   * For each request rate limiter whose type that is not of the type of the request which got rejected,
+   * check if slot borrowing is enabled. If enabled, try to acquire a slot.
+   * If allotted, return else try next request type.
+   */
+  private RequestRateLimiter trySlotBorrowing(String requestType) {
+    for (Map.Entry<String, RequestRateLimiter> currentEntry : requestRateLimiterMap.entrySet()) {
+      RequestRateLimiter requestRateLimiter = currentEntry.getValue();
+
+      if (requestRateLimiter.getRateLimiterConfig().requestType.equals(requestType)) {
+        continue;
+      }
+
+      if (requestRateLimiter.getRateLimiterConfig().isSlotBorrowingEnabled && requestRateLimiter.allowSlotBorrowing()) {
+        return requestRateLimiter;
+      }
+    }
+
+    return null;
+  }
+
+  // Decrement the active requests in the rate limiter for the corresponding request type.
+  public void decrementActiveRequests(HttpServletRequest request) {
```

Comment:
I feel like this is insufficient. If this request is part of its own queue, but there is another request of this same type that is currently borrowing a slot from another queue, then we shout prioritize returning the borrowed slot instead of our own.

This might have potential performance implications and can lead to livelock/request starvation.

Reply:
Discussed offline. Marking experimental and will follow up with a JIRA for rethinking the same.